### PR TITLE
SALTO-2547: Serialization by context for automations in Jira

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Add a `GetLookupNameFunc` in Jira to enable serialization by context

---

References in Jira automations can be serialized by id or by name, for example:
image

We want to be able to determine the serialization method during deploy, this change adds a `GetLookupNameFunc` that resolves the value according to its type.

---

Release Notes:

Jira

- More references will be added to Jira automation

---

User Notifications:

Jira

- More references will be added to Jira automation

### CODE REVIEW

File: packages/adapter-utils/src/utils.ts [hunk 1]
```diff
@@ -463,6 +464,7 @@ export const resolveValues: ResolveValuesFunc = async (
         ) : value,
         field,
         path,
+        element,
```
Comment (ori-moisis) [inline packages/adapter-utils/src/utils.ts:467]: I have not read this code and did not review the rest, but looking at the infra changes - is this not duplication with the `topLevelParent` that we have inside the reference?
why do we need both?

Comment (shir-reifenberg) [reply packages/adapter-utils/src/utils.ts:467]: Cause in this case we want the source element that contains the reference and not the target element

File: packages/jira-adapter/src/reference_mapping.ts [hunk 2]
```diff
@@ -444,20 +444,53 @@ ReferenceContextStrategyName
     serializationStrategy: 'id',
     target: { typeContext: 'parentSelectedFieldType' },
   },
+  {
```
Comment (netama) [inline packages/jira-adapter/src/reference_mapping.ts:447]: nit: let's add a comment here as well next to each pair of rules explaining that they overlap and that the strategy for serialization is determined by `serializationMethodLookUp`

File: packages/jira-adapter/src/reference_mapping.ts [hunk 3]
```diff
@@ -444,20 +444,53 @@ ReferenceContextStrategyName
...
+  },
   {
     src: { field: 'value', parentTypes: [AUTOMATION_FIELD] },
     serializationStrategy: 'id',
     target: { typeContext: 'parentFieldType' },
   },
 ]
 
+// during deploy, serialization method for references
+// from automations is determined by automations values
```
Comment (netama) [inline packages/jira-adapter/src/reference_mapping.ts:470]: 1. comment styling nit:
```suggestion
/**
 * during deploy, serialization method for references
 * from automations is determined by automations values
 */
```
2. I might be missing something but I'm not sure how to read this sentence - maybe something like 
> Determine serialization strategy for references with overlapping serialization rules in automation instances: value / values fields will be resolved based on the neighbor type field
3. naming nit: this name is a little too generic, maybe something like `getAutomationValuesLookupFunc`

File: packages/jira-adapter/src/filters/automation/automation_name_references.ts [hunk 4]
```diff
@@ -1,53 +0,0 @@
-/*
```
Comment (netama) [inline packages/jira-adapter/src/filters/automation/automation_name_references.ts:1]: is this related to this change?

**Review by netama — COMMENTED** [review]
looks great 💯 added a few minor questions and suggestions

File: packages/jira-adapter/src/filters/automation/automation_name_references.ts (hunk 4 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/automation/automation_name_references.ts:1]: Yep. To handle the double serialization before this, we created a filter that changes the reference method from 'NAME' to 'ID'. Now that we can use both serialization methods its irrelevant

File: packages/jira-adapter/src/reference_mapping.ts [hunk 5]
```diff
@@ -466,9 +470,11 @@ ReferenceContextStrategyName
   },
 ]
 
-// during deploy, serialization method for references
-// from automations is determined by automations values
-export const serializationMethodLookUp: GetLookupNameFunc = ({
+/**
+* Determine serialization strategy for references with overlapping serialization rules
+* in automation instances: relevant fields will be resolved based on the neighbor type field
+*/
```
Comment (netama) [inline packages/jira-adapter/src/reference_mapping.ts:476]:
```suggestion
/**
 * Determine serialization strategy for references with overlapping serialization rules
 * in automation instances: relevant fields will be resolved based on the neighbor type field
 */
```

File: packages/jira-adapter/src/filters/automation/automation_name_references.ts (hunk 4 above)
Comment (netama) [reply packages/jira-adapter/src/filters/automation/automation_name_references.ts:1]: I'm not fully following this one - let's discuss IRL?

Comment (netama) [reply packages/jira-adapter/src/filters/automation/automation_name_references.ts:1]: discussed IRL (thanks!) - this was a hack that relied on the fact that jira could accept both name and id during deploy, so we were "ok" with changing the value and adjusted the type so that it matches a different rule. now we're removing the hack

**Review by netama — APPROVED** [review]
